### PR TITLE
[templates] Update templates for the next SDK

### PR DESCRIPTION
**Pull request by Kudo**
# Why

follow up with #14347 

# How

from #14347 
- `et update-project-templates`
- manually updated some dependencies like `react`, `react-dom`, `react-native`, `react-native-web`, `@types/react`
- migrated bare template from unimodules to expo modules

new
- [android] migrate unimodules to expo-modules
- [android] integrate ReactNativeHostWrapper / ReactActivityDelegateWrapper
- replace splash screen integration by in-module setup
- replace expo-updates integration by in-module setup
- add Podfile.properties.json with hermes support
- Bump deployment target to 12.0
- fix other `node_modules` resolution where breaking monorepo support, e.g. hermes-engine or maven repo. 
- use mavenCentral before jcenter because [jcenter will be a read-only mirror repository](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/)

## Prerequisite
- [x] revert #14350 and publish new expo-modules-core without broken xcframework
- [x] publish new expo-modules-autolinking with #14384. otherwise no swift runtime will be generated as no any swift files by default.
- [x] #14438
- [x] disable prebuilding expo-updates 

 Co-authored-by: Tomasz Sapeta <tomasz.sapeta@swmansion.com>
 Co-authored-by: Cedric van Putten <me@bycedric.com>

## TODO
- [ ] if #14428 landed, add `EXAppDefines` integration

# Test Plan

```sh
expo init -t /path/to/templates/expo-template-bare-minimum sdk43
cd sdk43
expo run:android
expo run:android --variant release
expo run:ios
expo run:ios --configuration Release
```

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by brentvatne — APPROVED** [review]
this looks great!

one nit - on ios i'm seeing this warning on each app load:

>  WARN  No native splash screen registered for given view controller. Call 'SplashScreen.show' for given view controller first.

we can handle this in a follow up pr

**Comment by Kudo** [comment]
> > WARN  No native splash screen registered for given view controller. Call 'SplashScreen.show' for given view controller first.
> 
> we can handle this in a follow up pr

it's a regression of #14063. proposing the fix in: #14467

File: templates/expo-template-bare-minimum/ios/Podfile [hunk 1]
```diff
@@ -1,15 +1,20 @@
...
 require File.join(File.dirname(`node --print "require.resolve('react-native/package.json')"`), "scripts/react_native_pods")
-require File.join(File.dirname(`node --print "require.resolve('react-native-unimodules/package.json')"`), "cocoapods.rb")
-require File.join(File.dirname(`node --print "require.resolve('expo-modules-core/package.json')"`), "scripts/autolinking")
 require File.join(File.dirname(`node --print "require.resolve('@react-native-community/cli-platform-ios/package.json')"`), "native_modules")
 
-platform :ios, '11.0'
+platform :ios, '12.0'
+
+require 'json'
+podfile_properties = JSON.parse(File.read('./Podfile.properties.json')) rescue {}
```
Comment (tsapeta) [inline templates/expo-template-bare-minimum/ios/Podfile:8]: It might be better to have this in `package.json` — what do you think?

**Review by tsapeta — APPROVED** [review]

File: templates/expo-template-bare-minimum/ios/Podfile (hunk 1 above)
Comment (Kudo) [reply templates/expo-template-bare-minimum/ios/Podfile:8]: once thought of this. but config-plugin may change the target (in this case, package.json) during expo-cli prebuilding and makes git in dirty state. a separated `ios/Podfile.properties.json` could prevent this case. user who want to keep managed workflow could also safely remove entire ios/ folder.

**Comment by Kudo** [comment]
going to merge this first and leave #14467 as follow up pr accordingly.

**Pull request merged**